In [3]:
import pandas as pd
data=pd.read_csv('https://raw.githubusercontent.com/sumathi16/Machine-Learning-with-Python/master/data/global_sales_data/master.csv')
data.head()

Unnamed: 0    Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  \
0           0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01   
1           1  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00   
2           2  Ord_5446   Prod_6  SHP_7608  Cust_1818   164.02      0.03   
3           3  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01   
4           4  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09   

   Order_Quantity   Profit  Shipping_Cost             ...              \
0              23   -30.51           3.60             ...               
1              26  1148.90           2.50             ...               
2              23   -47.64           6.15             ...               
3              13     4.56           0.93             ...               
4              43   729.34          14.30             ...               

   Customer_Name Province Region  Customer_Segment Order_ID_y    Ship_Mode  \
0  AARON BERGMAN  ALBERTA   WEST         CORPORATE      36262  REGULAR AIR   
1  AARON BERGMAN  ALBERTA   WEST         CORPORATE      36262  EXPRESS AIR   
2  AARON BERGMAN  ALBERTA   WEST         CORPORATE      36262  EXPRESS AIR   
3  AARON BERGMAN  ALBERTA   WEST         CORPORATE      20513  EXPRESS AIR   
4  AARON BERGMAN  ALBERTA   WEST         CORPORATE      39682  EXPRESS AIR   

    Ship_Date  Ship_Month Ship_Year                    Duration  
0  2010-07-28           7      2010   1 days 00:00:00.000000000  
1  2010-07-27           7      2010   0 days 00:00:00.000000000  
2  2010-07-28           7      2010   1 days 00:00:00.000000000  
3  2009-08-07           8      2009  31 days 00:00:00.000000000  
4  2010-11-11          11      2010  61 days 00:00:00.000000000  

[5 rows x 28 columns]

In [4]:
data.columns

Index(['Unnamed: 0', 'Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales',
       'Discount', 'Order_Quantity', 'Profit', 'Shipping_Cost',
       'Product_Base_Margin', 'Product_Category', 'Product_Sub_Category',
       'Order_ID_x', 'Order_Date', 'Order_Priority', 'Order_Month',
       'Order_Year', 'Customer_Name', 'Province', 'Region', 'Customer_Segment',
       'Order_ID_y', 'Ship_Mode', 'Ship_Date', 'Ship_Month', 'Ship_Year',
       'Duration'],
      dtype='object')

In [6]:
data.drop('Unnamed: 0',axis=1,inplace=True)

KeyError: "['Unnamed: 0'] not found in axis"

In [7]:
data

Ord_id  Prod_id   Ship_id    Cust_id       Sales  Discount  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818    136.8100      0.01   
1     Ord_5446   Prod_4  SHP_7610  Cust_1818   4701.6900      0.00   
2     Ord_5446   Prod_6  SHP_7608  Cust_1818    164.0200      0.03   
3     Ord_5406  Prod_13  SHP_7549  Cust_1818     42.2700      0.01   
4     Ord_5456   Prod_6  SHP_7625  Cust_1818   2337.8900      0.09   
5     Ord_5485  Prod_17  SHP_7664  Cust_1818   4233.1500      0.08   
6     Ord_2978  Prod_16  SHP_4112  Cust_1088    305.0500      0.04   
7     Ord_5484  Prod_16  SHP_7663  Cust_1820    322.8200      0.05   
8     Ord_5430   Prod_4  SHP_7588  Cust_1820    337.2035      0.09   
9     Ord_5462   Prod_4  SHP_7632  Cust_1820    175.8565      0.03   
10    Ord_5462   Prod_5  SHP_7631  Cust_1820    323.9600      0.07   
11    Ord_5489  Prod_17  SHP_7671  Cust_1820    754.9200      0.00   
12    Ord_5481  Prod_15  SHP_7659  Cust_1820   2700.7800      0.06   
13    Ord_3730  Prod_16  SHP_5175  Cust_1314    459.0800      0.04   
14    Ord_3731  Prod_13  SHP_5176  Cust_1314    193.7000      0.06   
15    Ord_3733   Prod_4  SHP_5178  Cust_1314    553.3415      0.08   
16    Ord_4143  Prod_16  SHP_5771  Cust_1417    207.2100      0.06   
17    Ord_4143   Prod_6  SHP_5770  Cust_1417    901.3200      0.10   
18    Ord_4796  Prod_16  SHP_6686  Cust_1659     95.0900      0.09   
19    Ord_4796   Prod_6  SHP_6686  Cust_1659    122.0900      0.04   
20    Ord_4796  Prod_13  SHP_6687  Cust_1659    150.0600      0.02   
21    Ord_4830   Prod_6  SHP_6731  Cust_1659    158.9100      0.06   
22    Ord_1854  Prod_16  SHP_2548   Cust_727    182.4700      0.05   
23    Ord_1854   Prod_6  SHP_2548   Cust_622    817.5300      0.03   
24    Ord_1854  Prod_17  SHP_2548   Cust_564     25.9600      0.06   
25    Ord_1814  Prod_13  SHP_2507   Cust_622     81.9000      0.01   
26    Ord_1814   Prod_9  SHP_2507  Cust_1550    495.5000      0.06   
27    Ord_2957  Prod_13  SHP_4078   Cust_564     65.1200      0.07   
28    Ord_2957   Prod_6  SHP_4078   Cust_564    205.8700      0.03   
29    Ord_2991  Prod_13  SHP_4136   Cust_564    193.2500      0.03   
...        ...      ...       ...        ...         ...       ...   
8369  Ord_1035   Prod_1  SHP_1429   Cust_394   4754.0800      0.00   
8370  Ord_1035   Prod_8  SHP_1428   Cust_394    357.4300      0.08   
8371   Ord_994   Prod_1  SHP_1375   Cust_369  11103.7500      0.10   
8372  Ord_3944   Prod_1  SHP_5482  Cust_1363    294.9100      0.03   
8373  Ord_2987   Prod_8  SHP_4127  Cust_1101    990.7300      0.02   
8374  Ord_4538   Prod_8  SHP_6317  Cust_1552    105.1300      0.08   
8375  Ord_4388   Prod_8  SHP_6118  Cust_1501    448.2300      0.01   
8376  Ord_2114   Prod_8  SHP_2890   Cust_806    195.8300      0.04   
8377   Ord_464   Prod_8   SHP_623   Cust_165    638.7200      0.04   
8378  Ord_4566   Prod_8  SHP_6353  Cust_1515    676.6900      0.09   
8379  Ord_4566  Prod_14  SHP_6354  Cust_1515  13698.9600      0.05   
8380  Ord_4412   Prod_8  SHP_6149  Cust_1515    268.5500      0.10   
8381  Ord_2980   Prod_8  SHP_4114  Cust_1091   1854.9400      0.04   
8382  Ord_2995   Prod_8  SHP_4143  Cust_1113    677.9000      0.05   
8383  Ord_2992   Prod_8  SHP_4139  Cust_1110   1177.5000      0.02   
8384    Ord_24   Prod_8    SHP_32    Cust_19   1078.4900      0.08   
8385  Ord_1860   Prod_8  SHP_2554   Cust_654   2925.7800      0.03   
8386  Ord_1016   Prod_8  SHP_1406   Cust_383    138.0500      0.09   
8387  Ord_3676   Prod_8  SHP_5095  Cust_1294   1008.1900      0.03   
8388  Ord_3676   Prod_8  SHP_5095  Cust_1294   2440.2600      0.00   
8389  Ord_3665   Prod_8  SHP_5080  Cust_1294    401.1400      0.10   
8390   Ord_961   Prod_8  SHP_1330   Cust_350   2575.0100      0.09   
8391  Ord_4350   Prod_8  SHP_6065  Cust_1476     82.7200      0.08   
8392  Ord_4567   Prod_8  SHP_6355  Cust_1476    925.4300      0.00   
8393   Ord_943   Prod_8  SHP_1304   Cust_337     38.9900      0.09   
8394   Ord_973   P

### How much sales happening in each customer segment

In [8]:
data.Customer_Segment.nunique()

4

In [9]:
data.Customer_Segment.value_counts()

CORPORATE         3076
HOME OFFICE       2032
CONSUMER          1649
SMALL BUSINESS    1642
Name: Customer_Segment, dtype: int64

In [10]:
(data['Sales'][data.Customer_Segment=='CORPORATE']).sum()

5498904.8774999995

In [11]:
segments=data.Customer_Segment.value_counts().index
segments

Index(['CORPORATE', 'HOME OFFICE', 'CONSUMER', 'SMALL BUSINESS'], dtype='object')

In [12]:
Sales_Segments=[]
for segment in segments:
    Sales_Segments.append(
        (data['Sales'][data.Customer_Segment==segment]).sum())
Sales_Segments

[5498904.8774999995, 3564763.875, 3063611.082, 2788320.9895]

### Create a pandas series with values as sum in each category and index as segment name

In [13]:
Sales_Segments=pd.Series(Sales_Segments,segments)
Sales_Segments

CORPORATE         5.498905e+06
HOME OFFICE       3.564764e+06
CONSUMER          3.063611e+06
SMALL BUSINESS    2.788321e+06
dtype: float64

### Findout the segment name where we have the highest sales

In [14]:
Sales_Segments.idxmax()

'CORPORATE'

In [15]:
# Find the average sales in each customer segment
Sales_Segments=[]
for segment in segments:
    Sales_Segments.append(
        (data['Sales'][data.Customer_Segment==segment]).mean())
Sales_Segments

[1787.680389304291, 1754.3129306102326, 1857.8599648271663, 1698.1248413520116]

### GROUP BY 

In [16]:
data_cust=data.groupby('Customer_Segment')
data_cust.Sales.sum()

Customer_Segment
CONSUMER          3.063611e+06
CORPORATE         5.498905e+06
HOME OFFICE       3.564764e+06
SMALL BUSINESS    2.788321e+06
Name: Sales, dtype: float64

In [17]:
data.groupby('Customer_Segment').Sales.sum()

Customer_Segment
CONSUMER          3.063611e+06
CORPORATE         5.498905e+06
HOME OFFICE       3.564764e+06
SMALL BUSINESS    2.788321e+06
Name: Sales, dtype: float64

In [18]:
data.groupby('Customer_Segment')['Sales'].sum()

Customer_Segment
CONSUMER          3.063611e+06
CORPORATE         5.498905e+06
HOME OFFICE       3.564764e+06
SMALL BUSINESS    2.788321e+06
Name: Sales, dtype: float64

In [19]:
data_cust.Sales.sum().idxmax()

'CORPORATE'

In [20]:
data_cust.Sales.sum().argmax()

C:\Users\sivar\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 'argmax' is deprecated, use 'idxmax' instead. The behavior of 'argmax'
will be corrected to return the positional maximum in the future.
Use 'series.values.argmax' to get the position of the maximum now.
  """Entry point for launching an IPython kernel.


'CORPORATE'

### Minimum Sales average in which customer segment

In [21]:
data_cust.Sales.mean().idxmin()

'SMALL BUSINESS'

In [22]:
data_cust.Sales.mean()

Customer_Segment
CONSUMER          1857.859965
CORPORATE         1787.680389
HOME OFFICE       1754.312931
SMALL BUSINESS    1698.124841
Name: Sales, dtype: float64

### Maximum Sales average in which customer segment

In [23]:
data_cust.Sales.mean().idxmax()

'CONSUMER'

In [24]:
data_cust.mean()

Sales  Discount  Order_Quantity      Profit  \
Customer_Segment                                                      
CONSUMER          1857.859965  0.049903       25.324439  174.627010   
CORPORATE         1787.680389  0.049841       25.525683  194.975943   
HOME OFFICE       1754.312931  0.049444       25.987697  156.670290   
SMALL BUSINESS    1698.124841  0.049403       25.391596  192.270408   

                  Shipping_Cost  Product_Base_Margin    Order_ID_x  \
Customer_Segment                                                     
CONSUMER              13.024748             0.512905  29651.300788   
CORPORATE             12.698911             0.512783  29526.969766   
HOME OFFICE           12.771757             0.513608  30227.655020   
SMALL BUSINESS        12.995840             0.510258  30776.489647   

                  Order_Month   Order_Year    Order_ID_y  Ship_Month  \
Customer_Segment                                                       
CONSUMER             6.588842  2010.428745  29651.300788    6.548211   
CORPORATE            6.560143  2010.500325  29526.969766    6.561118   
HOME OFFICE          6.528051  2010.504429  30227.655020    6.553150   
SMALL BUSINESS       6.709501  2010.476248  30776.489647    6.584653   

                    Ship_Year  
Customer_Segment               
CONSUMER          2010.434203  
CORPORATE         2010.505202  
HOME OFFICE       2010.507874  
SMALL BUSINESS    2010.482948

In [24]:
data_cust.describe()

Discount                                                   \
                    count      mean       std  min   25%   50%   75%   max   
Customer_Segment                                                             
CONSUMER           1649.0  0.049903  0.031653  0.0  0.02  0.05  0.08  0.11   
CORPORATE          3076.0  0.049841  0.031751  0.0  0.02  0.05  0.08  0.21   
HOME OFFICE        2032.0  0.049444  0.031782  0.0  0.02  0.05  0.08  0.25   
SMALL BUSINESS     1642.0  0.049403  0.032202  0.0  0.02  0.05  0.08  0.17   

                 Order_ID_x                 ...   Ship_Year          \
                      count          mean   ...         75%     max   
Customer_Segment                            ...                       
CONSUMER             1649.0  29651.300788   ...      2011.0  2012.0   
CORPORATE            3076.0  29526.969766   ...      2012.0  2012.0   
HOME OFFICE          2032.0  30227.655020   ...      2012.0  2012.0   
SMALL BUSINESS       1642.0  30776.489647   ...      2011.0  2012.0   

                 Shipping_Cost                                           \
                         count       mean        std   min   25%    50%   
Customer_Segment                                                          
CONSUMER                1649.0  13.024748  17.576852  0.49  2.99  6.170   
CORPORATE               3076.0  12.698911  17.114333  0.49  3.50  6.035   
HOME OFFICE             2032.0  12.771757  17.308675  0.49  3.30  6.070   
SMALL BUSINESS          1642.0  12.995840  17.183757  0.49  2.99  6.130   

                                   
                      75%     max  
Customer_Segment                   
CONSUMER          13.9900  154.12  
CORPORATE         13.6600  143.71  
HOME OFFICE       13.9900  164.73  
SMALL BUSINESS    14.2725  147.12  

[4 rows x 96 columns]

### In which  product-sub category we are getting minimum profit

In [25]:
data.groupby('Product_Sub_Category').Profit.sum()

Product_Sub_Category
APPLIANCES                         97158.06
BINDERS AND BINDER ACCESSORIES    307413.39
BOOKCASES                         -33582.13
CHAIRS & CHAIRMATS                149649.73
COMPUTER PERIPHERALS               94287.48
COPIERS AND FAX                   167361.49
ENVELOPES                          48182.60
LABELS                             13677.17
OFFICE FURNISHINGS                100427.93
OFFICE MACHINES                   307712.93
PAPER                              45263.20
PENS & ART SUPPLIES                 7564.78
RUBBER BANDS                        -102.67
SCISSORS, RULERS AND TRIMMERS      -7799.25
STORAGE & ORGANIZATION              6664.15
TABLES                            -99062.50
TELEPHONES AND COMMUNICATION      316951.62
Name: Profit, dtype: float64

In [26]:
data.groupby('Product_Sub_Category').Profit.sum().idxmin()

'TABLES'

In [27]:
data.groupby(['Order_Year','Order_Month']).Sales.sum()

Order_Year  Order_Month
2009        1              379695.0180
            2              336525.2050
            3              411076.4110
            4              439218.8355
            5              382541.0940
            6              286397.2530
            7              339817.0355
            8              301470.8270
            9              307395.8935
            10             346021.8880
            11             265888.1155
            12             413091.8795
2010        1              240197.9760
            2              243504.4415
            3              273429.5910
            4              242681.2995
            5              272234.3790
            6              279740.5190
            7              283801.7100
            8              212134.3055
            9              463269.2195
            10             350759.6650
            11             323332.1120
            12             364595.5865
2011        1              191642.3765
 

In [28]:
data.groupby(['Order_Year','Order_Month']).Sales.sum().min()

183279.70799999996

In [29]:
data.groupby(['Order_Year','Order_Month']).Sales.sum().idxmin()

(2012, 8)

In [30]:
data.groupby(['Product_Category','Product_Sub_Category']).sum()

Sales  Discount  \
Product_Category Product_Sub_Category                                    
FURNITURE        BOOKCASES                        822652.040      9.09   
                 CHAIRS & CHAIRMATS              1761836.550     19.42   
                 OFFICE FURNISHINGS               698093.810     38.46   
                 TABLES                          1896008.142     18.00   
OFFICE SUPPLIES  APPLIANCES                       736991.540     21.98   
                 BINDERS AND BINDER ACCESSORIES  1022957.590     46.10   
                 ENVELOPES                        174085.800     12.02   
                 LABELS                            38981.550     14.27   
                 PAPER                            446452.860     60.53   
                 PENS & ART SUPPLIES              167107.220     32.46   
                 RUBBER BANDS                      15006.630      9.22   
                 SCISSORS, RULERS AND TRIMMERS     80996.310      7.09   
                 STORAGE & ORGANIZATION          1070182.600     27.89   
TECHNOLOGY       COMPUTER PERIPHERALS             795875.940     36.82   
                 COPIERS AND FAX                 1130361.300      4.48   
                 OFFICE MACHINES                 2168697.140     17.02   
                 TELEPHONES AND COMMUNICATION    1889313.802     42.34   

                                                 Order_Quantity     Profit  \
Product_Category Product_Sub_Category                                        
FURNITURE        BOOKCASES                                 4891  -33582.13   
                 CHAIRS & CHAIRMATS                        9590  149649.73   
                 OFFICE FURNISHINGS                       20648  100427.93   
                 TABLES                                    9195  -99062.50   
OFFICE SUPPLIES  APPLIANCES                               10901   97158.06   
                 BINDERS AND BINDER ACCESSORIES           22992  307413.39   
                 ENVELOPES                                 6578   48182.60   
                 LABELS                                    7301   13677.17   
                 PAPER                                    30871   45263.20   
                 PENS & ART SUPPLIES                      16980    7564.78   
                 RUBBER BANDS                              4920    -102.67   
                 SCISSORS, RULERS AND TRIMMERS             3931   -7799.25   
                 STORAGE & ORGANIZATION                   13804    6664.15   
TECHNOLOGY       COMPUTER PERIPHERALS                     19222   94287.48   
                 COPIERS AND FAX                           2128  167361.49   
                 OFFICE MACHINES                           7856  307712.93   
                 TELEPHONES AND COMMUNICATION             22969  316951.62   

                                                 Shipping_Cost  \
Product_Category Product_Sub_Category                            
FURNITURE        BOOKCASES                             8646.07   
                 CHAIRS & CHAIRMATS                   15512.69   
                 OFFICE FURNISHINGS                    8402.72   
                 TABLES                               20682.21   
OFFICE SUPPLIES  APPLIANCES                            6854.11   
                 BINDERS AND BINDER ACCESSORIES        6633.52   
                 ENVELOPES                             1682.77   
                 LABELS                                 288.66   
                 PAPER                                 7914.41   
                 PENS & ART SUPPLIES                   2041.81   
                 RUBBER BANDS                           225.21   
                 SCISSORS, RULERS AND TRIMMERS          670.51   
                 STORAGE & ORGANIZATION                9784.51   
TECHNOLOGY       COMPUTER PERIPHERALS                  4067.34   
                 COPIERS AND FAX                       2446.88   
                 OFFICE MACHINES           

In [31]:
data.groupby(['Product_Category','Product_Sub_Category'])\
[['Profit','Sales']].sum()

Profit        Sales
Product_Category Product_Sub_Category                                  
FURNITURE        BOOKCASES                       -33582.13   822652.040
                 CHAIRS & CHAIRMATS              149649.73  1761836.550
                 OFFICE FURNISHINGS              100427.93   698093.810
                 TABLES                          -99062.50  1896008.142
OFFICE SUPPLIES  APPLIANCES                       97158.06   736991.540
                 BINDERS AND BINDER ACCESSORIES  307413.39  1022957.590
                 ENVELOPES                        48182.60   174085.800
                 LABELS                           13677.17    38981.550
                 PAPER                            45263.20   446452.860
                 PENS & ART SUPPLIES               7564.78   167107.220
                 RUBBER BANDS                      -102.67    15006.630
                 SCISSORS, RULERS AND TRIMMERS    -7799.25    80996.310
                 STORAGE & ORGANIZATION            6664.15  1070182.600
TECHNOLOGY       COMPUTER PERIPHERALS             94287.48   795875.940
                 COPIERS AND FAX                 167361.49  1130361.300
                 OFFICE MACHINES                 307712.93  2168697.140
                 TELEPHONES AND COMMUNICATION    316951.62  1889313.802

- Create a new column is_profitable with values True or False by checking  whether the value is +ve or -ve
- Update theprofit column values with single decimal point
- Ex: 30.51 to -30.5

In [33]:
data['Profit']

0        -30.51
1       1148.90
2        -47.64
3          4.56
4        729.34
5       1219.87
6         23.12
7        -17.58
8       -120.93
9        -93.26
10      -136.82
11      -129.57
12      -793.36
13        61.57
14       -14.99
15       -67.71
16       -78.64
17       338.01
18       -13.53
19       -15.20
20        30.97
21       -38.57
22        -3.18
23        76.63
24       -29.29
25         4.06
26       -12.67
27         1.42
28        65.88
29        71.55
         ...   
8369     299.88
8370    -135.98
8371    -522.59
8372     -37.27
8373      56.39
8374    -103.23
8375     110.60
8376      10.79
8377    -130.88
8378     138.10
8379   -9078.94
8380     -99.02
8381    -559.78
8382     196.36
8383     420.05
8384     252.66
8385    1041.29
8386    -100.49
8387     210.01
8388     917.28
8389    -131.03
8390     231.46
8391     -29.71
8392     368.86
8393     -33.52
8394     -20.90
8395     -31.38
8396      21.68
8397    2787.59
8398     196.08
Name: Profit, Length: 83

In [34]:
data['is_profitable']=data['Profit']>0
data.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin',
       'Product_Category', 'Product_Sub_Category', 'Order_ID_x', 'Order_Date',
       'Order_Priority', 'Order_Month', 'Order_Year', 'Customer_Name',
       'Province', 'Region', 'Customer_Segment', 'Order_ID_y', 'Ship_Mode',
       'Ship_Date', 'Ship_Month', 'Ship_Year', 'Duration', 'is_profitable'],
      dtype='object')

In [35]:
data['is_profitable'][:5]

0    False
1     True
2    False
3     True
4     True
Name: is_profitable, dtype: bool

In [36]:
def is_profitable(x):
    return x>0

In [37]:
data['Profit'].apply(is_profitable)

0       False
1        True
2       False
3        True
4        True
5        True
6        True
7       False
8       False
9       False
10      False
11      False
12      False
13       True
14      False
15      False
16      False
17       True
18      False
19      False
20       True
21      False
22      False
23       True
24      False
25       True
26      False
27       True
28       True
29       True
        ...  
8369     True
8370    False
8371    False
8372    False
8373     True
8374    False
8375     True
8376     True
8377    False
8378     True
8379    False
8380    False
8381    False
8382     True
8383     True
8384     True
8385     True
8386    False
8387     True
8388     True
8389    False
8390     True
8391    False
8392     True
8393    False
8394    False
8395    False
8396     True
8397     True
8398     True
Name: Profit, Length: 8399, dtype: bool

### Update the profit column values with single decimal point
- Ex: -30.51 to -30.05

In [38]:
def number(x):
    return round(x,1)
print(number(54.27897))
print(number(53.926754))

54.3
53.9


In [39]:
data['Profit']=data['Profit'].apply(number)
data['Profit']

0        -30.5
1       1148.9
2        -47.6
3          4.6
4        729.3
5       1219.9
6         23.1
7        -17.6
8       -120.9
9        -93.3
10      -136.8
11      -129.6
12      -793.4
13        61.6
14       -15.0
15       -67.7
16       -78.6
17       338.0
18       -13.5
19       -15.2
20        31.0
21       -38.6
22        -3.2
23        76.6
24       -29.3
25         4.1
26       -12.7
27         1.4
28        65.9
29        71.5
         ...  
8369     299.9
8370    -136.0
8371    -522.6
8372     -37.3
8373      56.4
8374    -103.2
8375     110.6
8376      10.8
8377    -130.9
8378     138.1
8379   -9078.9
8380     -99.0
8381    -559.8
8382     196.4
8383     420.1
8384     252.7
8385    1041.3
8386    -100.5
8387     210.0
8388     917.3
8389    -131.0
8390     231.5
8391     -29.7
8392     368.9
8393     -33.5
8394     -20.9
8395     -31.4
8396      21.7
8397    2787.6
8398     196.1
Name: Profit, Length: 8399, dtype: float64

In [40]:
data['Profit'].apply(lambda x:round(x,1))

0        -30.5
1       1148.9
2        -47.6
3          4.6
4        729.3
5       1219.9
6         23.1
7        -17.6
8       -120.9
9        -93.3
10      -136.8
11      -129.6
12      -793.4
13        61.6
14       -15.0
15       -67.7
16       -78.6
17       338.0
18       -13.5
19       -15.2
20        31.0
21       -38.6
22        -3.2
23        76.6
24       -29.3
25         4.1
26       -12.7
27         1.4
28        65.9
29        71.5
         ...  
8369     299.9
8370    -136.0
8371    -522.6
8372     -37.3
8373      56.4
8374    -103.2
8375     110.6
8376      10.8
8377    -130.9
8378     138.1
8379   -9078.9
8380     -99.0
8381    -559.8
8382     196.4
8383     420.1
8384     252.7
8385    1041.3
8386    -100.5
8387     210.0
8388     917.3
8389    -131.0
8390     231.5
8391     -29.7
8392     368.9
8393     -33.5
8394     -20.9
8395     -31.4
8396      21.7
8397    2787.6
8398     196.1
Name: Profit, Length: 8399, dtype: float64

In [41]:
help(pd.DataFrame.pivot_table)

Help on function pivot_table in module pandas.core.frame:

pivot_table(self, values=None, index=None, columns=None, aggfunc='mean', fill_value=None, margins=False, dropna=True, margins_name='All')
    Create a spreadsheet-style pivot table as a DataFrame. The levels in
    the pivot table will be stored in MultiIndex objects (hierarchical
    indexes) on the index and columns of the result DataFrame
    
    Parameters
    ----------
    values : column to aggregate, optional
    index : column, Grouper, array, or list of the previous
        If an array is passed, it must be the same length as the data. The
        list can contain any of the other types (except list).
        Keys to group by on the pivot table index.  If an array is passed,
        it is being used as the same manner as column values.
    columns : column, Grouper, array, or list of the previous
        If an array is passed, it must be the same length as the data. The
        list can contain any of the other types

In [42]:
data.pivot_table(values='Sales',index='Customer_Segment',columns='Product_Category')

Product_Category    FURNITURE  OFFICE SUPPLIES   TECHNOLOGY
Customer_Segment                                           
CONSUMER          3153.092777       799.285815  2918.830136
CORPORATE         2942.876104       798.402161  3007.534303
HOME OFFICE       3165.876645       838.475467  2742.959557
SMALL BUSINESS    2757.176869       826.097641  2852.441325

In [43]:
data.pivot_table(index='Customer_Segment')

Discount    Order_ID_x    Order_ID_y  Order_Month  \
Customer_Segment                                                      
CONSUMER          0.049903  29651.300788  29651.300788     6.588842   
CORPORATE         0.049841  29526.969766  29526.969766     6.560143   
HOME OFFICE       0.049444  30227.655020  30227.655020     6.528051   
SMALL BUSINESS    0.049403  30776.489647  30776.489647     6.709501   

                  Order_Quantity   Order_Year  Product_Base_Margin  \
Customer_Segment                                                     
CONSUMER               25.324439  2010.428745             0.512905   
CORPORATE              25.525683  2010.500325             0.512783   
HOME OFFICE            25.987697  2010.504429             0.513608   
SMALL BUSINESS         25.391596  2010.476248             0.510258   

                      Profit        Sales  Ship_Month    Ship_Year  \
Customer_Segment                                                     
CONSUMER          174.627653  1857.859965    6.548211  2010.434203   
CORPORATE         194.976658  1787.680389    6.561118  2010.505202   
HOME OFFICE       156.670669  1754.312931    6.553150  2010.507874   
SMALL BUSINESS    192.270646  1698.124841    6.584653  2010.482948   

                  Shipping_Cost  is_profitable  
Customer_Segment                                
CONSUMER              13.024748       0.500910  
CORPORATE             12.698911       0.481469  
HOME OFFICE           12.771757       0.498524  
SMALL BUSINESS        12.995840       0.496346

In [44]:
import numpy as np
data.pivot_table(index='Customer_Segment',aggfunc=np.sum)

Discount  Order_ID_x  Order_ID_y  Order_Month  \
Customer_Segment                                                  
CONSUMER             82.29    48894995    48894995        10865   
CORPORATE           153.31    90824959    90824959        20179   
HOME OFFICE         100.47    61422595    61422595        13265   
SMALL BUSINESS       81.12    50534996    50534996        11017   

                  Order_Quantity  Order_Year  Product_Base_Margin    Profit  \
Customer_Segment                                                              
CONSUMER                   41760     3315197               838.60  287961.0   
CORPORATE                  78517     6184299              1566.04  599748.2   
HOME OFFICE                52807     4085345              1036.46  318354.8   
SMALL BUSINESS             41693     3301202               831.21  315708.4   

                         Sales  Ship_Month  Ship_Year  Shipping_Cost  \
Customer_Segment                                                       
CONSUMER          3.063611e+06       10798    3315206       21477.81   
CORPORATE         5.498905e+06       20182    6184314       39061.85   
HOME OFFICE       3.564764e+06       13316    4085352       25952.21   
SMALL BUSINESS    2.788321e+06       10812    3301213       21339.17   

                  is_profitable  
Customer_Segment                 
CONSUMER                  826.0  
CORPORATE                1481.0  
HOME OFFICE              1013.0  
SMALL BUSINESS            815.0

In [45]:
rep=data['Product_Base_Margin'].mean()
rep

0.5125131957773537

In [46]:
data.pivot_table(values="Product_Base_Margin",index="Customer_Segment",aggfunc=np.sum,fill_value=100)

Product_Base_Margin
Customer_Segment                     
CONSUMER                       838.60
CORPORATE                     1566.04
HOME OFFICE                   1036.46
SMALL BUSINESS                 831.21

In [47]:
data['Customer_Segment'][0]='Unknown'
data['Product_Base_Margin'][0]=np.nan

C:\Users\sivar\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\sivar\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [48]:
data.iloc[0,:]

Ord_id                                       Ord_5446
Prod_id                                       Prod_16
Ship_id                                      SHP_7609
Cust_id                                     Cust_1818
Sales                                          136.81
Discount                                         0.01
Order_Quantity                                     23
Profit                                          -30.5
Shipping_Cost                                     3.6
Product_Base_Margin                               NaN
Product_Category                      OFFICE SUPPLIES
Product_Sub_Category    SCISSORS, RULERS AND TRIMMERS
Order_ID_x                                      36262
Order_Date                                 2010-07-27
Order_Priority                          NOT SPECIFIED
Order_Month                                         7
Order_Year                                       2010
Customer_Name                           AARON BERGMAN
Province                    

In [49]:
data.pivot_table(values='Product_Base_Margin',
                 index='Customer_Segment',aggfunc=np.sum,fill_value=45)

Product_Base_Margin
Customer_Segment                     
CONSUMER                       838.60
CORPORATE                     1565.48
HOME OFFICE                   1036.46
SMALL BUSINESS                 831.21
Unknown                          0.00

In [50]:
data.pivot_table(values='Product_Base_Margin',
                 index='Customer_Segment',aggfunc=np.sum)

Product_Base_Margin
Customer_Segment                     
CONSUMER                       838.60
CORPORATE                     1565.48
HOME OFFICE                   1036.46
SMALL BUSINESS                 831.21
Unknown                          0.00

In [51]:
data.pivot_table(values=['Product_Base_Margin','Profit'],
                 index='Customer_Segment',aggfunc=np.sum)

Product_Base_Margin    Profit
Customer_Segment                               
CONSUMER                       838.60  287961.0
CORPORATE                     1565.48  599778.7
HOME OFFICE                   1036.46  318354.8
SMALL BUSINESS                 831.21  315708.4
Unknown                          0.00     -30.5

In [52]:
data.pivot_table(values=['Product_Base_Margin','Profit'],
                 index='Customer_Segment',
                 aggfunc={'Product_Base_Margin':np.max,'Profit':np.mean})

Product_Base_Margin      Profit
Customer_Segment                                 
CONSUMER                         0.85  174.627653
CORPORATE                        0.85  195.049984
HOME OFFICE                      0.85  156.670669
SMALL BUSINESS                   0.85  192.270646
Unknown                           NaN  -30.500000

### Get the categorial column names
- data.dtypes
- data.columns
- fancy indexing

In [53]:
data.dtypes

Ord_id                   object
Prod_id                  object
Ship_id                  object
Cust_id                  object
Sales                   float64
Discount                float64
Order_Quantity            int64
Profit                  float64
Shipping_Cost           float64
Product_Base_Margin     float64
Product_Category         object
Product_Sub_Category     object
Order_ID_x                int64
Order_Date               object
Order_Priority           object
Order_Month               int64
Order_Year                int64
Customer_Name            object
Province                 object
Region                   object
Customer_Segment         object
Order_ID_y                int64
Ship_Mode                object
Ship_Date                object
Ship_Month                int64
Ship_Year                 int64
Duration                 object
is_profitable              bool
dtype: object

In [54]:
type(data.dtypes)

pandas.core.series.Series

In [55]:
data.dtypes.index

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin',
       'Product_Category', 'Product_Sub_Category', 'Order_ID_x', 'Order_Date',
       'Order_Priority', 'Order_Month', 'Order_Year', 'Customer_Name',
       'Province', 'Region', 'Customer_Segment', 'Order_ID_y', 'Ship_Mode',
       'Ship_Date', 'Ship_Month', 'Ship_Year', 'Duration', 'is_profitable'],
      dtype='object')

In [56]:
data.dtypes== object

Ord_id                   True
Prod_id                  True
Ship_id                  True
Cust_id                  True
Sales                   False
Discount                False
Order_Quantity          False
Profit                  False
Shipping_Cost           False
Product_Base_Margin     False
Product_Category         True
Product_Sub_Category     True
Order_ID_x              False
Order_Date               True
Order_Priority           True
Order_Month             False
Order_Year              False
Customer_Name            True
Province                 True
Region                   True
Customer_Segment         True
Order_ID_y              False
Ship_Mode                True
Ship_Date                True
Ship_Month              False
Ship_Year               False
Duration                 True
is_profitable           False
dtype: bool

In [57]:
data.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin',
       'Product_Category', 'Product_Sub_Category', 'Order_ID_x', 'Order_Date',
       'Order_Priority', 'Order_Month', 'Order_Year', 'Customer_Name',
       'Province', 'Region', 'Customer_Segment', 'Order_ID_y', 'Ship_Mode',
       'Ship_Date', 'Ship_Month', 'Ship_Year', 'Duration', 'is_profitable'],
      dtype='object')

In [58]:
type(data.columns)

pandas.core.indexes.base.Index

In [59]:
# categorical columns
data.columns[data.dtypes==object]

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Product_Category',
       'Product_Sub_Category', 'Order_Date', 'Order_Priority', 'Customer_Name',
       'Province', 'Region', 'Customer_Segment', 'Ship_Mode', 'Ship_Date',
       'Duration'],
      dtype='object')

In [60]:
data.columns[data.dtypes==np.int64]

Index(['Order_Quantity', 'Order_ID_x', 'Order_Month', 'Order_Year',
       'Order_ID_y', 'Ship_Month', 'Ship_Year'],
      dtype='object')

In [61]:
type(data.describe())

pandas.core.frame.DataFrame

In [62]:
data.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184951      12.838557   
std     3585.050525     0.031823       14.481071   1196.653385      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.300000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.700000     164.730000   

       Product_Base_Margin    Order_ID_x  Order_Month   Order_Year  \
count          8335.000000   8399.000000  8399.000000  8399.000000   
mean              0.512507  29965.179783     6.587213  2010.482557   
std               0.135596  17260.883447     3.424515     1.123860   
min               0.350000      3.000000     1.000000  2009.000000   
25%               0.380000  15011.500000     4.000000  2009.000000   
50%               0.520000  29857.000000     7.000000  2010.000000   
75%               0.590000  44596.000000    10.000000  2012.000000   
max               0.850000  59973.000000    12.000000  2012.000000   

         Order_ID_y   Ship_Month    Ship_Year  
count   8399.000000  8399.000000  8399.000000  
mean   29965.179783     6.561257  2010.487558  
std    17260.883447     3.433550     1.123715  
min        3.000000     1.000000  2009.000000  
25%    15011.500000     4.000000  2009.000000  
50%    29857.000000     7.000000  2010.000000  
75%    44596.000000     9.000000  2012.000000  
max    59973.000000    12.000000  2012.000000

In [63]:
data.describe(include=object).columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Product_Category',
       'Product_Sub_Category', 'Order_Date', 'Order_Priority', 'Customer_Name',
       'Province', 'Region', 'Customer_Segment', 'Ship_Mode', 'Ship_Date',
       'Duration'],
      dtype='object')

In [64]:
data.Order_Priority.value_counts()

HIGH             1768
LOW              1720
NOT SPECIFIED    1672
MEDIUM           1631
CRITICAL         1608
Name: Order_Priority, dtype: int64

- NOT SPECIFIED-0
- LOW-1
- MEDIUM-2
- HIGH-3
- CRITICAL-4

In [67]:
data['Order_Priority_Number']

KeyError: 'Order_Priority_Number'

Define a function
- input is a string
- return a value must be 0,1,2,3 or 4

In [68]:
def str_to_num(x):
    if x=='NOT SPECIFIED':
        return 0
    elif x=='LOW':
        return 1
    elif x=='MEDIUM':
        return 2
    elif x=='HIGH':
        return 3
    elif x=='CRITICAL':
        return 4

In [69]:
data['Order_Priority_Number']=data['Order_Priority'].apply(str_to_num)
data['Order_Priority_Number'].value_counts()

3    1768
1    1720
0    1672
2    1631
4    1608
Name: Order_Priority_Number, dtype: int64

In [70]:
data.Order_Priority_Number.value_counts()

3    1768
1    1720
0    1672
2    1631
4    1608
Name: Order_Priority_Number, dtype: int64

In [71]:
data.Order_Priority[10:30]

10    NOT SPECIFIED
11             HIGH
12         CRITICAL
13           MEDIUM
14           MEDIUM
15             HIGH
16         CRITICAL
17         CRITICAL
18         CRITICAL
19         CRITICAL
20         CRITICAL
21             HIGH
22         CRITICAL
23         CRITICAL
24         CRITICAL
25    NOT SPECIFIED
26    NOT SPECIFIED
27         CRITICAL
28         CRITICAL
29              LOW
Name: Order_Priority, dtype: object

### Creating dummy columns

In [72]:
import pandas as pd
pd.get_dummies(data['Order_Priority'])

CRITICAL  HIGH  LOW  MEDIUM  NOT SPECIFIED
0            0     0    0       0              1
1            0     0    0       0              1
2            0     0    0       0              1
3            0     1    0       0              0
4            0     0    0       1              0
5            0     0    1       0              0
6            0     1    0       0              0
7            0     0    1       0              0
8            0     0    0       0              1
9            0     0    0       0              1
10           0     0    0       0              1
11           0     1    0       0              0
12           1     0    0       0              0
13           0     0    0       1              0
14           0     0    0       1              0
15           0     1    0       0              0
16           1     0    0       0              0
17           1     0    0       0              0
18           1     0    0       0              0
19           1     0    0       0              0
20           1     0    0       0              0
21           0     1    0       0              0
22           1     0    0       0              0
23           1     0    0       0              0
24           1     0    0       0              0
25           0     0    0       0              1
26           0     0    0       0              1
27           1     0    0       0              0
28           1     0    0       0              0
29           0     0    1       0              0
...        ...   ...  ...     ...            ...
8369         0     1    0       0              0
8370         0     1    0       0              0
8371         1     0    0       0              0
8372         0     1    0       0              0
8373         0     0    0       0              1
8374         0     0    0       1              0
8375         0     0    1       0              0
8376         0     0    1       0              0
8377         0     1    0       0              0
8378         0     0    0       0              1
8379         0     0    0       0              1
8380         0     0    0       1              0
8381         0     0    0       0              1
8382         0     0    1       0              0
8383         0     1    0       0              0
8384         0     0    1       0              0
8385         0     0    0       0              1
8386         0     0    1       0              0
8387         0     0    0       1              0
8388         0     0    0       1              0
8389         0     0    0       1              0
8390         0     0    1       0              0
8391         0     0    1       0              0
8392         0     0    0       0              1
8393         0     1    0       0              0
8394         0     0    0       1              0
8395         0     1    0       0              0
8396         0     0    1       0              0
8397         0     0    0       0              1
8398         0     0    1       0              0

[8399 rows x 5 columns]

In [73]:
data1=pd.get_dummies(data)
data1

Sales  Discount  Order_Quantity  Profit  Shipping_Cost  \
0       136.8100      0.01              23   -30.5           3.60   
1      4701.6900      0.00              26  1148.9           2.50   
2       164.0200      0.03              23   -47.6           6.15   
3        42.2700      0.01              13     4.6           0.93   
4      2337.8900      0.09              43   729.3          14.30   
5      4233.1500      0.08              35  1219.9          26.30   
6       305.0500      0.04              27    23.1           3.37   
7       322.8200      0.05              35   -17.6           3.98   
8       337.2035      0.09               9  -120.9           4.99   
9       175.8565      0.03              10   -93.3           1.25   
10      323.9600      0.07              40  -136.8           7.96   
11      754.9200      0.00               7  -129.6          19.99   
12     2700.7800      0.06              29  -793.4          58.20   
13      459.0800      0.04              34    61.6           3.14   
14      193.7000      0.06              20   -15.0           4.82   
15      553.3415      0.08              33   -67.7           0.99   
16      207.2100      0.06              24   -78.6           6.14   
17      901.3200      0.10              25   338.0           5.08   
18       95.0900      0.09               9   -13.5           3.37   
19      122.0900      0.04               6   -15.2           9.54   
20      150.0600      0.02              49    31.0           0.70   
21      158.9100      0.06              24   -38.6           5.20   
22      182.4700      0.05              17    -3.2           3.37   
23      817.5300      0.03              34    76.6          11.54   
24       25.9600      0.06               2   -29.3           6.24   
25       81.9000      0.01              27     4.1           0.96   
26      495.5000      0.06              44   -12.7           6.97   
27       65.1200      0.07              38     1.4           1.00   
28      205.8700      0.03              35    65.9           1.46   
29      193.2500      0.03              19    71.5           1.09   
...          ...       ...             ...     ...            ...   
8369   4754.0800      0.00              21   299.9          28.66   
8370    357.4300      0.08              17  -136.0           6.50   
8371  11103.7500      0.10              32  -522.6          99.00   
8372    294.9100      0.03              23   -37.3           6.27   
8373    990.7300      0.02              30    56.4           4.00   
8374    105.1300      0.08               4  -103.2           4.00   
8375    448.2300      0.01              26   110.6           4.00   
8376    195.8300      0.04              24    10.8           1.99   
8377    638.7200      0.04              20  -130.9           5.50   
8378    676.6900      0.09              21   138.1           1.99   
8379  13698.9600      0.05               4 -9078.9          24.49   
8380    268.5500      0.10              36   -99.0           2.83   
8381   1854.9400      0.04               6  -559.8           7.18   
8382    677.9000      0.05              20   196.4           1.99   
8383   1177.5000      0.02              34   420.1           1.99   
8384   1078.4900      0.08              27   252.7           1.99   
8385   2925.7800      0.03              29  1041.3           7.18   
8386    138.0500      0.09              19  -100.5           5.53   
8387   1008.1900      0.03              23   210.0           1.99   
8388   2440.2600      0.00              50   917.3           1.99   
8389    401.1400      0.10              10  -131.0           6.50   
8390   2575.0100      0.09              27   231.5          19.99   
8391     82.7200      0.08               6   -29.7           1.99   
8392    925.4300      0.00              28   368.9           1.99   
8393     38.9900      0.09              23   -33.5           1.99   
8394     29.0500      0.08               4   -20.9           4.32   
8395     

In [74]:
data1.shape[1]

18957

In [75]:
cat_cols=data.columns[data.dtypes==object]
cat_cols

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Product_Category',
       'Product_Sub_Category', 'Order_Date', 'Order_Priority', 'Customer_Name',
       'Province', 'Region', 'Customer_Segment', 'Ship_Mode', 'Ship_Date',
       'Duration'],
      dtype='object')

In [76]:
data[cat_cols].nunique().sum()

18943

In [77]:
189843-len(cat_cols)+data.shape[1]

189857